sdk/core/azure-core-v2/src/main/java/com/azure/v2/core/implementation/AccessTokenCache.java

```diff
@@ -71,69 +71,64 @@ public AccessToken getToken(TokenRequestContext tokenRequestContext, boolean che
+        return fetchToken(tokenRequestContext, forceFetchToken);
+    }
+
+    private void validateTokenRequestContext(TokenRequestContext tokenRequestContext) {
         if (tokenRequestContext == null) {
             throw LOGGER
                 .logThrowableAsError(new IllegalArgumentException("The token request context input cannot be null."));
         }
+    }
+
+    private AccessToken fetchToken(TokenRequestContext tokenRequestContext, boolean forceFetchToken) {
+        OffsetDateTime now = OffsetDateTime.now();
         AccessTokenCacheInfo cache = this.cacheInfo;
         AccessToken cachedToken = cache.getCachedAccessToken();
 
-        OffsetDateTime now = OffsetDateTime.now();
-        boolean tokenRefresh;
-        AccessToken fallback;
+        boolean needsRefresh = determineRefreshRequirement(tokenRequestContext, forceFetchToken, cachedToken, now);
 
-        // Check if the incoming token request context is different from the cached one. A different
-        // token request context, requires to fetch a new token as the cached one won't work for the
-        // passed in token request context.
-        boolean forceRefresh
-            = (forceFetchToken && checkIfForceRefreshRequired(tokenRequestContext)) || this.tokenRequestContext == null;
+        if (!needsRefresh && cachedToken != null) {
+            return cachedToken;
+        }
+
+        return attemptTokenRefresh(cachedToken, tokenRequestContext, now);
+    }
 
-        if (forceRefresh) {
+    private boolean determineRefreshRequirement(TokenRequestContext tokenRequestContext, boolean forceFetchToken,
+        AccessToken cachedToken, OffsetDateTime now) {
+        if (forceFetchToken && checkIfForceRefreshRequired(tokenRequestContext)) {
             this.tokenRequestContext = tokenRequestContext;
-            tokenRefresh = true;
-            fallback = null;
-        } else if (cachedToken != null && !checkIfRefreshIsNeeded(cachedToken)) {
-            // fresh cache & no need to refresh
-            tokenRefresh = false;
-            fallback = cachedToken;
-        } else if (cachedToken == null || cachedToken.isExpired()) {
-            // no token to use
-            // refresh immediately
-            tokenRefresh = true;
+            return true;
+        }
 
-            // cache doesn't exist or expired, no fallback
-            fallback = null;
-        } else {
-            // token available, but close to expiry
-            if (now.isAfter(cache.getNextTokenRefresh())) {
-                // refresh immediately
-                tokenRefresh = true;
-            } else {
-                // still in timeout, do not refresh
-                tokenRefresh = false;
-            }
-            // cache hasn't expired, ignore refresh error this time
-            fallback = cachedToken;
+        if (cachedToken == null || cachedToken.isExpired()) {
+            return true;
         }
 
+        if (!checkIfRefreshIsNeeded(cachedToken)) {
```

Comment:
why negate and return false here? Can't this just return true also?

Reply:
To ensure, we check the last scenario of respecting the REFRESH_DELAY, in the event this checkIfRefreshIsNeeded returns true.